### ISSUE TITLE
After some time kubernetes stops scheduling Pods all together

### ISSUE THREAD

**Issue by dvaldivia** [kubernetes-sigs/kind#1990]
**What happened**:

Often after some time has passed the kubernetes cluster inside Kind stops scheduling pods all together, the pod doesn't even show as `Pending` it simply doesn't show, the solution commonly is to delete the kind cluster and create it again.

**What you expected to happen**:

If I delete a pod it should be re-scheduled (if it's a statefulset or deployment)

**Anything else we need to know?**:

Also happens on macOS

**Environment:**

- kind version: (use `kind version`): kind v0.9.0 go1.15.6 linux/amd64
- Kubernetes version: (use `kubectl version`): client v1.19.4 , server v1.19.1
- Docker version: (use `docker info`): Server Version: 20.10.1
- OS (e.g. from `/etc/os-release`): 20.04.1 LTS (Focal Fossa)

**Comment by aojea** [comment]
you'll need to provide a bit more of information ... if you can reproduce it again, maybe you can upload the `kind export logs` folder

**Comment by dvaldivia** [comment]
Here are the logs

https://www.icloud.com/iclouddrive/0YMS6NdwhcXXzuUJFc7l_o3Fg#kind-logs

**Comment by dvaldivia** [comment]
@aojea I've tried to debug this before, but no logs in k8s explains why the pods are not scheduling at all

**Comment by aojea** [comment]
it seems that the control-plane node starts with one ip
```
2020-12-21T17:28:45.060060178Z stderr F Flag --insecure-port has been deprecated, This flag will be removed in a future version.
2020-12-21T17:28:45.060167589Z stderr F I1221 17:28:45.060076       1 server.go:625] external host was not specified, using 172.18.0.3
2020-12-21T17:28:45.060362011Z stderr F I1221 17:28:45.060312       1 server.go:163] Version: v1.19.1
```

and it changes later
```
2020-12-21T17:28:50.605318063Z stderr F I1221 17:28:50.605226       1 main.go:65] hostIP = 172.18.0.3
2020-12-21T17:28:50.605348674Z stderr F podIP = 172.18.0.5
2020-12-21T17:28:50.607894633Z stderr F panic: hostIP(= "172.18.0.3") != podIP(= "172.18.0.5") but must be running with host network: 
```

the controllers try to connect to the new IP
```
2020-12-23T18:26:46.88171023Z stderr F E1223 18:26:46.881498       1 reflector.go:127] k8s.io/client-go/informers/factory.go:134: Failed to watch *v1.Service: failed to list *v1.Service: Get "https://172.18.0.5:6443/api/v1/services?limit=500&resourceVersion=0": dial tcp 172.18.0.5:6443: connect: connection refused
``` 

but the apiserver is still listening in the 172.18.0.3

are you doing upgrades or restarting containers?

**Comment by dvaldivia** [comment]
sometimes I restart the computer, some other times I do `apt upgrade` so maybe that's what causing it.

Do you think there's a way to fix it without deleting the kind cluster?

**Comment by aojea** [comment]
restart should work,  ... `apt upgrade` better not doing it, I don't know what can happen ...

**Comment by dvaldivia** [comment]
Whenever I reach this hiatus state I've tried restarting the machine without much success, so perhaps it's a bad idea to do `apt upgrade`, I'll keep an eye and see if this happens again without `apt upgrade`.

Thanks for the help understanding this problem

**Comment by BenTheElder** [comment]
@aojea this seems like a kindnetd bug, shouldn't we just get the pod restarted?

**Comment by aojea** [comment]
> @aojea this seems like a kindnetd bug, shouldn't we just get the pod restarted?

kindnetd is working ok, it is restarts and works ok, I just copied the log to highlight that something is happening in the container, that is reporting different hostIP and PodIP, despite is a hostnetwork container ... is not kindnetd for sure ;)

**Comment by BenTheElder** [comment]
> kindnetd is working ok, it is restarts and works ok, I just copied the log to highlight that something is happening in the container, that is reporting different hostIP and PodIP, despite is a hostnetwork container ... is not kindnetd for sure ;)

can it retain the old container / pod IP when the host changes?
presumably if the pod failed and the sandbox were recreated it would become consistent again?

also: do we really need that check? we should probably just get the host IP and not care if pod IP is consistent? ... we don't support people running this image with arbitary manifests 🤷‍♂️

**Comment by aojea** [comment]
> can it retain the old container / pod IP when the host changes?

it seems so, honestly, the kubelet -> CRI -> CNI is a big drama right now ... there are some patches in kubelet addressing these inconsistencies in 1.21, at least in RedHat we have this as a TODO to fix soon 

> presumably if the pod failed and the sandbox were recreated it would become consistent again?

that is what happened, it paniced and when it restarted it was correct

> also: do we really need that check? we should probably just get the host IP and not care if pod IP is consistent? ... we don't support people running this image with arbitary manifests

I think we do need it, this saved us before and here too, the problem with removing the check is that you don't have any way to know that you are running wrong ... so kindnet maybe working but the cluster (or only some things in the cluster) not, and realising that this is caused because PodIP is != HostIP is not trivial

**Comment by BenTheElder** [comment]
I mean if we can detect they don't match and one of the two sources is correct just switch to exclusively using the correct source..?

**Comment by aojea** [comment]
> I mean if we can detect they don't match and one of the two sources is correct just switch to exclusively using the correct source..?

That is a bug in the kubelet , if is a host network pod, the pod IP and the host IP must be the same ... However, hostip is a long standing issue for sig-network, and it has to reviewed with the dual stack effort

**Comment by BenTheElder** [comment]
> That is a bug in the kubelet , if is a host network pod, the pod IP and the host IP must be the same ... However, hostip is a long standing issue for sig-network, and it has to reviewed with the dual stack effort


but kindnetd does not depend on actually having the IP for the pod? we just need to know what the IP is for generating the rules, if we can detect the real IP that's enough ...?

**Comment by aojea** [comment]
> but kindnetd does not depend on actually having the IP for the pod? we just need to know what the IP is for generating the rules, if we can detect the real IP that's enough ...?

yeah, we need the HostIP (that has to be the same as the NodeIP) ... the PodIP check is only to verify that is a pod using hostNetwork:true

**Comment by aojea** [comment]
> we can assume that though, since we control the podspec, and don't support kindnetd otherwise. Let's drop that check or just downgrade it to a warning log, and ensure we use the valid source of truth.

@BenTheElder I´ve downgraded it to a warning with the dual-stack changes
https://github.com/kubernetes-sigs/kind/pull/692/commits/791df83faaf1f3dfdee1d11c1127a8a1c337950b

**Comment by optimus-kart** [comment]
I am hitting this issue after restarting the host machine. is there any workaround for this?

**Comment by aojea** [comment]
/close
https://github.com/kubernetes-sigs/kind/commit/3dbeb894e3092a336ab4278d3823e73a1d66aff7#diff-02aa2d22d0befe557afb283f8925c8a006bb5f0318dc33a10176fc78299e87d6R109

it will be available in 0.11.0

**Comment by dvaldivia** [comment]
@aojea nice!
